src/artifacts.js

```diff
@@ -502,13 +502,27 @@ var replyWithArtifact = async function(taskId, runId, name, req, res) {
     return res.reportError('ResourceNotFound', 'Artifact not found', {});
   }
 
+  // Some downloading utilities need to know the artifact's storage type to be
+  // able to handle their downloads most correctly.  We're going to set this
+  // field on all artifact responses so that the downloading utilities can use
+  // slightly different logic for each artifact type
+  res.set('x-taskcluster-artifact-type', artifact.storageType);
```

Comment:
I would like to avoid this, it's rather implementation specific, and we have many clients.

I think we should document headers like `x-taskcluster-location-content-sha256`. For example we could say:

---
A taskcluster service MAY supply a `x-taskcluster-location-content-sha256` header when redirecting clients. If present clients SHOULD validate the content of the what is being downloaded from the `location` header against hexadecimal encoded sha256 from this header.

---

Or something like that... This way if we change the implementation, we can keep using the header... And if in the future we only use blob artifacts, we can change the documentation to say the header MUST be present, and SHOULD be validated.

Similarly, we can document other headers and how they should be handled. If we decide we need `sha3` in some future, it would be awesome that we don't have to invent a new artifact type. We just gracefully add it and later make it required. Or if we invent a new artifact type, then we can keep adding the `sha256` headers for compatibility with older clients; and they won't stop using it just because the implementation changed.

----
Note: we should probably write a page on the docs site where we explain these headers, what they are for and use words like MAY and SHOULD. Probably we could have a doc that says HTTP clients that use taskcluster APIs MUST follow redirects, and MUST NOT resend the `authentication` header when following a redirect. We haven't really documented this yet, we just sort of take it for granted -- pretty sure we agreed on things like this when jlal was around, but that's years ago.. maybe we should write it down?

As long as you keep reminding me that `transfer-sha256` headers are useful for supporting pass-through caches I think they are great. But maybe this is a good thing to write down in such docs as well.. As a side note that this is why it's useful :)

Reply:
Specifying the artifact *storage type* is not implementation specific at the level being suggested.  This is an implementation specific feature of the *entire* artifact subsystem, not of any specific type.  Specifying this value is important because certain clients (e.g. `taskcluster-lib-artifact{,-go}`) need to be able to tell the difference between the various artifact types.  When downloading an artifact, this library needs to know whether to do validation or not.  For a blob storage artifact, these headers are not an optional component.  They are 100% mandatory, so we should be validating that when we have a blob artifact, that they are present.  

Because we don't have these hash/size values for the S3 or Azure blob type, nor can we, we need to be able to ignore validation for those two artifact types.  For Error and Redirect artifacts, we would either need to start storing the SHA256 + Size of the response body at creation or when we send it.  How would we handle the validation of a redirect which can change?  We have different storage types, and each storage type might need slight different handling.  We could imply artifact type on response, but why?  We can easily be explicit without any detriment.

The differences between using Sha256 and a Sha-3 algorithm *can* be invisibly adapted within the blob artifact.  Which hashing algorithm used by a given storage type does not have *any* bearing on what artifact storage type it is.  It's either a blob, S3, Azure, Error or Redirect.  The change to Sha3 would represent such a large amount of work that changes to this header would be trivial, even if it were related.

I disagree that taskcluster api clients must follow redirects.  I think a more correct phrasing is that the get-artifact endpoint returns a response value which is in the form of a valid http 303 redirect with metadata exposed as X-Taskcluster prefixed headers.

One specific case that would violate this declaration is around HTTPS verification.  The `taskcluster-lib-artifact*` libraries understand that we should never follow redirects to non-https resources.  If we made it mandatory to follow the redirect and received a redirect to an HTTP resource, we'd still have to follow that redirect even if we didn't want the actual body data.

Another case is that someone is doing checks to see if an artifact on a test machine is valid.  The only way to get that information is to do a non-redirect-following call to get-artifact and find the data there.  If we made this mandatory, they'd have to redownload the whole thing just to get the Sha256 value.

Adding this header has literally zero impact on clients which follow the redirect.  It allows those clients which want to opt into certain behaviour the ability to do so explicitly instead of implicitly.  The value which this header represents is such a core concept to the artifact subsystem that any change it in would likely be a complete rewrite.

What are the specific technical concerns with exposing this information?

Reply:
> For a blob storage artifact, these headers are not an optional component. They are 100% mandatory, so we should be validating that when we have a blob artifact, that they are present.

We are largely in agreement, this is a technicality, @jhford let's talk on vidyo.... we can bike shed some other stuff too... I'll send an invite.

> We could imply artifact type on response, but why?

I wouldn't infer the artifact type; I would phrase it as clients should be unaware of the type. But that they SHOULD validate using headers like `x-taskcluster-location-content-sha256` if present.
So you get graceful degrading...

> Another case is that someone is doing checks to see if an artifact on a test machine is valid. 

Ideally that should use HEAD request.

> What are the specific technical concerns with exposing this information?

Tight coupling.... visibility of concepts that could be hidden. Yeah, this is all minor nits.

Reply:
okay, let's call it `x-taskcluster-storage-type`